### PR TITLE
(PUP-7079) Retry scheduled task deletion in specs

### PR DESCRIPTION
**Pull request by Iristyle**
- When run concurrently, there may be file sharing issues when using
   the task scheduler APIs that prevent a task deletion from
   succeeding with an error like:

   Failed to call
   with HRESULT: -2147024864.:  The process cannot access the file
   because it is being used by another process.

 - Introduce a task deletion helper that retries up to 3 times with
   a 1 second delay in between attempts, in an effort to clean up
   tasks under such circumstances

### CODE REVIEW

**Comment by Iristyle** [comment]
This is not awesome, but I don't see another great alternative here.

Note that we remoted into the machine that intermittently failed, and it *did* have a leftover scheduled task with a GUID based name... so aside from not writing tests that actually create tasks, a defensive delete is probably the best approach.

**Comment by MosesMendoza** [comment]
If we try and fail three times this will continue as success - previously it would fail. Is that desired? Not sure what the implications are of potentially leaving the tasks in place (other than modifying the test environment state)

**Comment by Iristyle** [comment]
@MosesMendoza the inability to delete a task is not indicative of an actual test failure... but it could mean that a task has been orphaned on a spec box.

The question is - is it important enough to fail a run, as a flag that someone should manually clean up the clutter later, in the event that 3 attempts fail?  A solo orphaned task on a machine shouldn't be a problem (as its uniquely named) - but if there are hundreds or thousands of orphans, I'm not sure if there is an impact.

FWIW, we've only seen this problem once... and when I remoted into the machine in question, it only had a single orphaned task in task scheduler.

**Comment by MosesMendoza** [comment]
So seems like probably not worth failing the run. In that case, could we rescue the specific exception that happens when the task delete fails? Or does it vary?

File: spec/unit/provider/scheduled_task/win32_taskscheduler_spec.rb [hunk 1]
```diff
@@ -1954,6 +1954,20 @@ def time_component
       end
     end
 
+    def delete_task_with_retry(name, attempts = 3)
+      failed = false
+
+      attempts.times do
+        begin
+          task.delete(name) if Win32::TaskScheduler.new.exists?(name)
+        rescue Puppet::Util::Windows::Error
```
Comment (Iristyle) [inline spec/unit/provider/scheduled_task/win32_taskscheduler_spec.rb]: Good suggestion on specifying an error type @MosesMendoza - updated ^

**Comment by MosesMendoza** [comment]
From the appveyor failure it looks like we might need to pass the task object into `delete_task_with_retry`

**Comment by Iristyle** [comment]
> From the appveyor failure it looks like we might need to pass the task object into `delete_task_with_retry`

Well, that's embarrassing... Wasn't it green before?

PR updated

**Comment by MosesMendoza** [comment]
Yeah I thought it was green too

**Comment by MosesMendoza** [comment]
👍  will merge when specs go green again

**Comment by MosesMendoza** [comment]
Since specs failed due to a reason this will ostensibly fix, going to go ahead and merge

**Pull request merged**